server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+          params.add(firstArgument);
+          params.add(secondArgument);
+          return new OperatorDto(UcdlToken.OR, params);
+        }
+        return firstArgument;
+      }
+      case "AND" -> {
+        AbstractSyntaxTreeDto firstArgument = buildAst(root.jjtGetChild(0), parameters);
+        if (root.jjtGetChild(1).jjtGetNumChildren() > 0) {
+          AbstractSyntaxTreeDto secondArgument =
+              buildAst(root.jjtGetChild(1).jjtGetChild(0), parameters);
+          List<AbstractSyntaxTreeDto> params = new ArrayList<>();
+          params.add(firstArgument);
+          params.add(secondArgument);
+          return new OperatorDto(UcdlToken.AND, params);
+        }
+        return firstArgument;
+      }
+      case "NOT" -> {
+        if (((SimpleNode) root).jjtGetFirstToken().image.equals("NOT")) {
+          List<AbstractSyntaxTreeDto> param = new ArrayList<>();
+          param.add(buildAst(root.jjtGetChild(0), parameters));
+          return new OperatorDto(UcdlToken.NOT, param);
+        }
+        return buildAst(root.jjtGetChild(0), parameters);
+      }
+      case "ISEMPTY" -> {
+        List<AbstractSyntaxTreeDto> param = new ArrayList<>();
+        param.add(buildAst(root.jjtGetChild(0), parameters));
+        return new OperatorDto(UcdlToken.IS_EMPTY, param);
+      }
+      case "ELEMENTINSETOREQUATION" -> {
+        List<AbstractSyntaxTreeDto> params = new ArrayList<>();
+        params.add(buildAst(root.jjtGetChild(0), parameters));
+
+        Node secondPart = root.jjtGetChild(1);
+
+        if (secondPart.toString().equals("ELEMENTINSET")) {
+          params.add(buildAst(secondPart.jjtGetChild(0), parameters));
+          return new OperatorDto(UcdlToken.IN, params);
+        }
+
+        if (secondPart.toString().equals("EQUATION")) {
+          params.add(buildAst(secondPart.jjtGetChild(1), parameters));
+          UcdlToken token = getComparatorToken(
+              ((SimpleNode) secondPart.jjtGetChild(0)).jjtGetFirstToken().image,
+              params
+          );
+          return new OperatorDto(token, params);
+        }
+        throw new IllegalStateException();
+      }
+      case "ELEMENT" -> {
+        if (root.jjtGetChild(0).toString().equals("NUMBERELEMENT")) {
+          return buildAst(root.jjtGetChild(0), parameters);
+        }
+
+        ValueDto<String> elementName = (ValueDto<String>) buildAst(root.jjtGetChild(0), parameters);
+
+        ResourceType elementType = parameters.get(elementName.value());
+
+        List<AbstractSyntaxTreeDto> attributes = new ArrayList<>();
+
+        Node attributeList = root.jjtGetChild(1);
+        while (attributeList.jjtGetNumChildren() > 0) {
+          String attribute = ((SimpleNode) attributeList).jjtGetFirstToken().next.image;
+          elementType = getNextResourceType(elementType, attribute);
+          //^ also performs semantic check whether the attribute is applicable to the given type
```

Comment:
Such comments are super helpful and imo they bring great value

Reply:
I guess that's sarcasm. But I can't tell for sure...

Reply:
No, I was actually serious lol. That's one of a few comments I actually really like :)